### PR TITLE
Trivial improvements for TapTree type

### PR DESCRIPTION
**Pull request by dr-orlovsky**
These are trivial fixes from extracted from now closed #922

### CODE REVIEW

File: src/util/psbt/map/output.rs [hunk 1]
```diff
@@ -125,6 +125,13 @@ impl TapTree {
         self.0
     }
 
+    /// Returns reference for the inner [`TaprootBuilder`]. The builder is guaranteed to be
+    /// finalized.
+    #[inline]
+    pub fn as_inner(&self) -> &TaprootBuilder {
```
Comment (apoelstra) [inline src/util/psbt/map/output.rs:131]: In f6dcd2cc56d942c33619bb526c089d171b04f9ab:

I think `as_inner' is an extremely weird name for this. Yes, as an implementation detail a `TapTree` is a finalized `TaprootBuilder`, but this doesn't necessarily make conceptual sense and I don't think we want to guarantee it.

I'd be ok with the same method named `as_builder`, or preferably `into_builder` which takes ownership.

File: src/util/psbt/map/output.rs [hunk 2]
```diff
@@ -92,33 +93,67 @@ impl PartialEq for TapTree {
...
 
-    /// Converts a [`TaprootBuilder`] into a tree if it is complete binary tree.
+    /// Constructs [`TapTree`] from a [`TaprootBuilder`] if it is complete binary tree.
     ///
     /// # Return
-    /// A `TapTree` iff the `inner` builder is complete, otherwise return the inner as `Err`.
-    pub fn from_inner(inner: TaprootBuilder) -> Result<Self, TaprootBuilder> {
-        if inner.is_complete() {
-            Ok(TapTree(inner))
+    /// A `TapTree` iff the `builder` builder is complete, otherwise return the inner as `Err`.
```
Comment (tcharding) [inline src/util/psbt/map/output.rs:136]:
```suggestion
    /// A `TapTree` iff the `builder` is complete, otherwise return the `builder` as `Err`.
```

Not part of this PR but the return section header should be plural i.e., `# Returns`.

File: src/util/psbt/map/output.rs [hunk 3]
```diff
@@ -93,6 +93,12 @@ impl PartialEq for TapTree {
 impl Eq for TapTree {}
 
 impl TapTree {
+    /// Convenience method to retrieve number of script leafs in the tree.
+    #[inline]
+    pub fn scripts_count(&self) -> usize {
```
Comment (sanket1729) [inline src/util/psbt/map/output.rs:120]: As mentioned elsewhere, this should implement `ExactLenIterator` instead

File: src/util/psbt/map/output.rs [hunk 4]
```diff
@@ -92,6 +92,13 @@ impl PartialEq for TapTree {
 
 impl Eq for TapTree {}
 
+impl AsRef<TaprootBuilder> for TapTree {
```
Comment (sanket1729) [inline src/util/psbt/map/output.rs:96]: nit: 

I would rather let users call `as_builder` instead rather than an indirect `AsRef`. Maybe it's just only me, but `TapTree` and `TaprootBuilder` are not similar conceptually in the operations they do. I would only conservatively implement `AsRef` when the conversion is super-obvious or should be hidden from the user.

File: src/util/psbt/map/output.rs [hunk 5]
```diff
@@ -99,6 +100,13 @@ impl AsRef<TaprootBuilder> for TapTree {
     }
 }
 
+impl Borrow<TaprootBuilder> for TapTree {
```
Comment (sanket1729) [inline src/util/psbt/map/output.rs:103]: Same comment as the previous commit. Is there any reason to do this? Have we done this for inner types before? I think we should avoid implementing it unless we have a reason. Ref, Deref (IMO) should only be implemented on types that are super-obvious to convert and only if they are really needed. 

Allowing function args that accept these is however another story.

**Review by sanket1729 — COMMENTED** [review]
One major comment.

File: src/util/psbt/map/output.rs (hunk 4 above)
Comment (dr-orlovsky) [reply src/util/psbt/map/output.rs:96]: `AsRef` is needed for generic function arguments

File: src/util/psbt/map/output.rs (hunk 5 above)
Comment (dr-orlovsky) [reply src/util/psbt/map/output.rs:103]: There is a reason to do this and I always do that for the new types. `Deref`, otherwise, is not always useful for them.

Some information: https://itfanr.gitbooks.io/rust-doc-en/content/borrow-and-asref.html

File: src/util/psbt/map/output.rs (hunk 3 above)
Comment (dr-orlovsky) [reply src/util/psbt/map/output.rs:120]: Sorry, this happens in #924: https://github.com/rust-bitcoin/rust-bitcoin/pull/924/commits/b5470bf568b93d124e2cadfaddc7bbe651d5304e

File: src/util/psbt/map/output.rs (hunk 4 above)
Comment (Kixunil) [reply src/util/psbt/map/output.rs:96]: I don't like leaking the internal representation at all. Is there a strong reason to do it?

File: src/util/psbt/map/output.rs [hunk 6]
```diff
@@ -92,33 +93,67 @@ impl PartialEq for TapTree {
...
+        &self.0
+    }
+}
+
+impl From<TapTree> for TaprootBuilder {
+    #[inline]
+    fn from(tree: TapTree) -> Self {
+        tree.into_builder()
+    }
+}
```
Comment (Kixunil) [inline src/util/psbt/map/output.rs:115]: Note that as opposed to previous conversion, this one is OK.

File: src/util/psbt/map/output.rs (hunk 3 above)
Comment (Kixunil) [reply src/util/psbt/map/output.rs:120]: `leaf_count` may be clearer name?

Comment (dr-orlovsky) [reply src/util/psbt/map/output.rs:120]: It is not leaf count: in the future with hidden parts of the tree we will not know about how many leaves are hidden. So probably better to have it as `known_scripts_count`.

File: src/util/psbt/map/output.rs (hunk 4 above)
Comment (dr-orlovsky) [reply src/util/psbt/map/output.rs:96]: Yes, otherwise we can't use `TapTree` in places where `TaprootBuilder` is required as an argument: `fn some_stuff(tree: impl AsRef<TaprootBuilder>)`

**Review by sanket1729 — COMMENTED** [review]
Had an incomplete review in another tab

File: src/util/psbt/map/output.rs (hunk 4 above)
Comment (sanket1729) [reply src/util/psbt/map/output.rs:96]: As mentioned in the above comment, I would rather users use conversion functions.

**Comment by sanket1729** [comment]
I have moved this to 0.29 since this is not a bug fix and a slight feature improvement with some bugs. There are some discussions that might take some time to resolve.  

At this rate, we will never have a release until we have a perfect taproot release(which is basically impossible). It has already been months since we started the RC cycle that we might have had two major releases.

**Comment by sanket1729** [comment]
@dr-orlovsky, do you have an objection to the 0.29 tag for this?

File: src/util/psbt/map/output.rs (hunk 2 above)
Comment (dr-orlovsky) [reply src/util/psbt/map/output.rs:136]: Fixed elsewhere: https://github.com/rust-bitcoin/rust-bitcoin/pull/929/files#diff-8ed6474e7f75e74be2ab10a3102be1af7dcda692d83edceff3012296e3a7e04eR140

**Comment by dr-orlovsky** [comment]
Removed everything you suggested to remove, @Kixunil and @sanket1729. I will not object to postpone it. But if you have time, I think it will be cool to get the rest of the pr (which is really trivial) just not to require breaking changes after.

**Review by sanket1729 — DISMISSED** [review]
utACK bcd9555ab096c3345c02d8d506abdbb11a867277.

**Comment by tcharding** [comment]
FWIW, looks good to me.

File: src/util/psbt/map/output.rs [hunk 7]
```diff
@@ -92,33 +92,47 @@ impl PartialEq for TapTree {
...
+    pub fn into_builder(self) -> TaprootBuilder {
         self.0
     }
 
+    /// Returns reference for the [`TaprootBuilder`] stored internally. The builder is guaranteed
+    /// to be finalized.
+    #[inline]
+    pub fn as_builder(&self) -> &TaprootBuilder {
+        &self.0
+    }
```
Comment (Kixunil) [inline src/util/psbt/map/output.rs:134]: You forgot to remove this one which also leaks internal representation.

**Review by Kixunil — CHANGES_REQUESTED** [review]
Just one detail, strong ACK renaming.

File: src/util/psbt/map/output.rs (hunk 7 above)
Comment (dr-orlovsky) [reply src/util/psbt/map/output.rs:134]: It can't be used for modifying the tree structure and is required to avoid a lot of clones (like `tap_tree.clone().into_builder()` downstream.

Comment (Kixunil) [reply src/util/psbt/map/output.rs:134]: Does it cause significant performance issues? Perhaps there's a more specific pattern of operations you often perform and could be handled differently?

Comment (Kixunil) [reply src/util/psbt/map/output.rs:134]: Actually, looking at `TaprootBuilder` API it doesn't accept `&self` in any modifying method so you'd still have to write `tap_tree.as_builder().clone()` and there's no perf difference. If the issue is just annoyance, I'm open to:

```rust
pub fn to_builder(&self) -> TaprootBuilder {
    self.clone().into_builder()
}
```

Comment (dr-orlovsky) [reply src/util/psbt/map/output.rs:133]: Did that

**Review by Kixunil — DISMISSED** [review]
ACK 798c08ece96308489174df8b5af69c2d055adc0b

**Review by tcharding — DISMISSED** [review]
ACK (excluding `into_` vs `to_`, I admit total defeat in trying to fully understand the naming conventions for as/into/to).

**Comment by tcharding** [comment]
> excluding `into_` vs `to_`, I admit total defeat in trying to fully understand the naming conventions for as/into/to

I rescind my statement of confusion, as if this minute I am in total understanding of that damnable table and ACK the naming `into_builder` and `to_builder` :)

**Comment by tcharding** [comment]
lolz, this one has `rc-fix` and milestone 0.29 - lets totally get rid of the rc-fix label soon as we do this release :)

**Comment by dr-orlovsky** [comment]
There was a conflict with the master; I have rebased. @sanket1729 @tcharding pls re-ACK

**Review by sanket1729 — APPROVED** [review]
ACK 4cdff06b1e54bd16b39626aa8eda8b3bd31e6e2c

**Review by apoelstra — APPROVED** [review]
ACK 4cdff06b1e54bd16b39626aa8eda8b3bd31e6e2c
